### PR TITLE
[query][lowering] lower TableWrite with native reader

### PR DESCRIPTION
**Pull request by catoverdrive**
This implementation lowers TableWrite with a TableNativeReader.

Punting on the `stageLocally` path for now (it'll throw a lowering error) since our current implementation adds a task listener to the spark task to clean up files, and I'm not sure how we want to handle that in the general case.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/TableWriter.scala [hunk 1]
```diff
@@ -101,7 +148,111 @@ case class PartitionNativeWriter(spec: AbstractTypedCodecSpec, partPrefix: Strin
     eltType: PStruct,
     mb: EmitMethodBuilder[_],
     region: Value[Region],
-    stream: SizedStream): EmitCode = ???
+    stream: SizedStream): EmitCode = {
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/expr/ir/TableWriter.scala:151]: Feel free to push back, this looks good, but is there any reason to not use the `scoped` methods for EmitCodeBuilder for a lot of this since we have the necessary EmitMethodBuilder to create the EmitCodeBuilder

File: hail/src/main/scala/is/hail/io/index/IndexWriter.scala [hunk 2]
```diff
@@ -380,4 +386,38 @@ class StagedIndexWriter(branchingFactor: Int, keyType: PType, annotationType: PT
     internalBuilder.create(cb)
     internalBuilder.store(cb, utils, 0)
   }
+
+  class NonImperativeIndexWriter(newInit: EmitMethodBuilder[_], newAdd: EmitMethodBuilder[_], newClose: EmitMethodBuilder[_]) {
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/io/index/IndexWriter.scala:390]: Why is this necessary? It seems like we can create a code builder and use the other version of the index writer.

File: hail/src/main/scala/is/hail/expr/ir/TableWriter.scala (hunk 1 above)
Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/ir/TableWriter.scala:151]: No, good call. I forgot that that existed. I'll refactor this.
